File: java-to-go-projects/18-pipeline/Stages.java
package pipeline;

import java.util.function.Function;
import java.util.function.Predicate;

/**
 * Collection of reusable pipeline stage functions.
 * Each stage reads from an input Channel and writes to an output Channel.
 *
 * Stages run in their own thread, processing items as they arrive.
 */
public class Stages {

    /**
     * Map stage: transforms each input item using the given function.
     */
    public static <T, R> Runnable map(Channel<T> in, Channel<R> out, Function<T, R> fn) {
        return () -> {
            try {
                while (true) {
                    T item = in.receive();
                    if (item == null && in.isClosed()) break;
                    if (item == null) continue;
                    out.send(fn.apply(item));
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            } finally {
                out.close();
            }
        };
    }

    /**
     * Filter stage: only passes items that match the predicate.
     */
    public static <T> Runnable filter(Channel<T> in, Channel<T> out, Predicate<T> pred) {
        return () -> {
            try {
                while (true) {
                    T item = in.receive();
                    if (item == null && in.isClosed()) break;
                    if (item == null) continue;
                    if (pred.test(item)) {
                        out.send(item);
                    }
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            } finally {
                out.close();
            }
        };
    }

    /**
     * Sink stage: collects all items into a list (terminal stage).
     */
    public static <T> Runnable sink(Channel<T> in, java.util.List<T> results) {
        return () -> {
            try {
                while (true) {
                    T item = in.receive();
                    if (item == null && in.isClosed()) break;
                    if (item == null) continue;
                    synchronized (results) {
                        results.add(item);
                    }
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        };
    }

    /**
     * Fan-out: duplicate input to multiple output channels.
     */
    public static <T> Runnable fanOut(Channel<T> in, java.util.List<Channel<T>> outs) {
        return () -> {
            try {
                while (true) {
                    T item = in.receive();
                    if (item == null && in.isClosed()) break;
                    if (item == null) continue;
                    for (Channel<T> out : outs) {
                        out.send(item);
                    }
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            } finally {
                for (Channel<T> out : outs) out.close();
            }
        };
    }
}

File: java-to-go-projects/18-pipeline/rust/src/stages.rs
use std::sync::mpsc::{Receiver, SyncSender};
use std::sync::{Arc, Mutex};
use std::thread::JoinHandle;

/// Spawn a thread that reads from rx, applies f to each item, sends result to tx.
/// Closes tx (drops) when rx is exhausted.
pub fn map_stage<T, U>(
    rx: Receiver<T>,
    tx: SyncSender<U>,
    f: impl Fn(T) -> U + Send + 'static,
) -> JoinHandle<()>
where
    T: Send + 'static,
    U: Send + 'static,
{
    todo!()
}

/// Spawn a thread that reads from rx, keeps items where predicate is true, sends to tx.
pub fn filter_stage<T>(
    rx: Receiver<T>,
    tx: SyncSender<T>,
    predicate: impl Fn(&T) -> bool + Send + 'static,
) -> JoinHandle<()>
where
    T: Send + 'static,
{
    todo!()
}

/// Spawn a thread that reads from rx and collects all items into results vec.
pub fn sink_stage<T>(
    rx: Receiver<T>,
    results: Arc<Mutex<Vec<T>>>,
) -> JoinHandle<()>
where
    T: Send + 'static,
{
    todo!()
}

/// Spawn a thread that reads from rx and sends each item to all output senders (fan-out).
pub fn fan_out<T>(
    rx: Receiver<T>,
    txs: Vec<SyncSender<T>>,
) -> JoinHandle<()>
where
    T: Send + Clone + 'static,
{
    todo!()
}
